File: ast/VAST.cpp
#include "VAST.h"

using json = nlohmann::json;
using namespace std;

namespace vast {
  VStatementExpr::VStatementExpr() {
    // Placeholder only
  }
  json VStatementExpr::toJson() {
    return "";
  }

  VImportList::VImportList(vector<VImport *> _imports) {
    imports = _imports;
  }

  json VImportList::toJson() {
    vector<json> importsJson;
    for(VImport *i : imports) {
      importsJson.push_back(i->toJson());
    }
    json importListJson = {
      {"ntype", "VImportList"}, {"imports", importsJson}
    };
    return importListJson;
  }

  VImport::VImport(string _path) {
    path = _path;
  }

  json VImport::toJson() {
    json importJson = {
      {"ntype", "VImport"}, {"path", path}
    };
    return importJson;
  }

  VBehavioralSpec::VBehavioralSpec(VImportList *_imports, VVarDeclList *_var_decs, VStatementExpr *_pre, VStatementExpr *_post) {
    imports = _imports;
    var_decs = _var_decs;
    pre = _pre;
    post = _post;
  }
  json VBehavioralSpec::toJson() {
    json varsJson = {};
    if (var_decs != nullptr) varsJson = var_decs->toJson();
    json preJson = {};
    if (pre != nullptr) preJson = pre->toJson();
    json postJson = post->toJson();
    json importsJson = imports->toJson();
    json total = {
      {"ntype", "VBehavioralSpec"}, {"imports", importsJson}, {"var_decls", varsJson}, {"pre", preJson}, {"post", postJson}
    };
    return total;
  }

  VTestSpec::VTestSpec(VImportList *_imports, VVarDeclList *_var_decs, VStatementExpr *_init, VStatementExpr *_spec) {
    imports = _imports;
    var_decs = _var_decs;
    init = _init;
    spec = _spec;
  }
  json VTestSpec::toJson() {
    json varsJson = {};
    if (var_decs != nullptr) varsJson = var_decs->toJson();
    json initJson = {};
    if (init != nullptr) initJson = init->toJson();
    json specJson = spec->toJson();
    json importsJson = imports->toJson();
    json total = {
      {"ntype", "VTestSpec"}, {"imports", importsJson}, {"var_decls", varsJson}, {"init", initJson}, {"spec", specJson}
    };
    return total;
  }

  VTempSpec::VTempSpec(VImportList *_imports, VVarDeclList *_var_decs, VStatementExpr *_fairness, VStatementExpr *_spec) {
    imports = _imports;
    var_decs = _var_decs;
    fairness = _fairness;
    spec = _spec;
  }
  json VTempSpec::toJson() {
    json varsJson = {};
    if (var_decs != nullptr) varsJson = var_decs->toJson();
    json fairnessJson = {};
    if (fairness != nullptr) fairnessJson = fairness->toJson();
    json specJson = spec->toJson();
    json importsJson = imports->toJson();
    json total = {
      {"ntype", "VTempSpec"}, {"imports", importsJson}, {"var_decls", varsJson}, {"fairness", fairnessJson}, {"spec", specJson}
    };
    return total;
  }

  VInvSpec::VInvSpec(VImportList *_imports, VVarDeclList *_var_decs, VStatementExpr *_inv) {
    imports = _imports;
    var_decs = _var_decs;
    inv = _inv;
  }
  json VInvSpec::toJson() {
    json varsJson = {};
    if (var_decs != nullptr) varsJson = var_decs->toJson();
    json invJson = inv->toJson();
    json importsJson = imports->toJson();
    json total = {
      {"ntype", "VInvSpec"}, {"imports", importsJson}, {"var_decls", varsJson}, {"inv", invJson}
    };
    return total;
  }

  VSynthSpec::VSynthSpec(VImportList *_imports, VVarDeclList *_var_decls, VStatementExpr *_init, VStatementExpr *_synth) {
    imports = _imports;
    var_decls = _var_decls;
    init = _init;
    synth = _synth;
  }
  json VSynthSpec::toJson() {
    json varsJson = {}; 
    if (var_decls != nullptr) varsJson = var_decls->toJson();
    json initJson = {}; 
    if (init != nullptr) initJson = init->toJson();
    json synthJson = synth->toJson();
    json importsJson = imports->toJson();
    json total = {
      {"ntype", "VSynthSpec"}, {"imports", importsJson}, {"var_decls", varsJson}, {"init", initJson}, {"synth", synthJson}
    };
    return total;
  }

  VVarDeclList::VVarDeclList(vector<VVarDecl*> _var_decs) {
    var_decs = _var_decs;
  }
  json VVarDeclList::toJson() {
    vector<json> varDecsJson;
    for (VVarDecl* v : var_decs) {
      varDecsJson.push_back(v->toJson());
    }
    json declJson = {
      {"ntype", "VVarDeclList"}, {"var_decls", varDecsJson}
    };
    return declJson;
  }

  VVarDecl::VVarDecl(VType* _typ, VID* _var) {
    typ = _typ;
    var = _var;
  }
  json VVarDecl::toJson() {
    json declJson = {
      {"ntype", "VVarDecl"}, {"type", typ->toJson()}, {"var", var->toJson()}
    };
    return declJson;
  }

  VID::VID(string _name) {
    name = _name;
  }
  json VID::toJson() {
    json vidJson = {
      {"ntype", "VID"}, {"name", name}
    };
    return vidJson;
  }

  VType::VType(string _name, bool _is_arr) {
    name = _name;
    is_arr = _is_arr;
  }
  json VType::toJson() {
    string typ_str = name;
    if (is_arr) typ_str = name + "[]";
    json vtypJson = {
      {"ntype", "VType"}, {"name", typ_str}
    };
    return vtypJson;
  }

  VBinStatementExpr::VBinStatementExpr(VStatementExpr *_lhs, VStatementExpr *_rhs, VBinOp *_op) {
    lhs = _lhs;
    rhs = _rhs;
    op = _op;
  }
  json VBinStatementExpr::toJson() {
    json lhsJson = lhs->toJson();
    json rhsJson = rhs->toJson();
    json vBinStmtExprJson = {
      {"ntype", "VBinStatementExpr"}, {"lhs", lhsJson}, {"rhs", rhsJson}, {"op", op->op}
    };
    return vBinStmtExprJson;
  }

  VBinOp::VBinOp(string _op) {
    op = _op;
  }
  json VBinOp::toJson() {
    json vBinOpJson = {
      {"ntype", "VBinOp"}, {"op", op}
    };
    return vBinOpJson;
  }

  VUnStatementExpr::VUnStatementExpr(VStatementExpr *_con, VUnOp *_op) {
    con = _con;
    op = _op;
  }
  json VUnStatementExpr::toJson() {
    json exprJson = con->toJson();
    json vUnStmtExprJson = {
      {"ntype", "VUnStatementExpr"}, {"expr", exprJson}, {"op", op->op}
    };
    return vUnStmtExprJson;
  }

  VUnOp::VUnOp(string _op) {
    op = _op;
  }
  json VUnOp::toJson() {
    json vUnOpJson = {
      {"ntype", "VUnOp"}, {"op", op}
    };
    return vUnOpJson;
  }

  VConstraintExpr::VConstraintExpr() {
    // Placeholder only
  }
  json VConstraintExpr::toJson() {
    return "";
  }

  VExecutedStatement::VExecutedStatement(VFunctionID *_fun, VConstraintExpr *_pre, VConstraintExpr *_con) {
    fun = _fun;
    pre = _pre;
    con = _con;
  }
  json VExecutedStatement::toJson() {
    json funJson = fun->toJson();
    json preJson = {};
    if (pre != nullptr) preJson = pre->toJson();
    json conJson = {};
    if (con != nullptr) conJson = con->toJson();
    json vExecutedStmtJson = {
      {"ntype", "VExecutedStatementExpr"}, {"fun", funJson}, {"pre", preJson}, {"con", conJson}
    };
    return vExecutedStmtJson;
  }

  VFinishedStatement::VFinishedStatement(VFunctionID *_fun, VConstraintExpr *_pre, VConstraintExpr *_con) {
    fun = _fun;
    pre = _pre;
    con = _con;
  }
  json VFinishedStatement::toJson() {
    json funJson = fun->toJson();
    json preJson = {};
    if (pre != nullptr) preJson = pre->toJson();
    json conJson = {};
    if (con != nullptr) conJson = con->toJson();
    json vFinishedStmtJson = {
      {"ntype", "VFinishedStatementExpr"}, {"fun", funJson}, {"pre", preJson}, {"con", conJson}
    };
    return vFinishedStmtJson;
  }

  VStartedStatement::VStartedStatement(VFunctionID *_fun, VConstraintExpr *_con) {
    fun = _fun;
    con = _con;
  }
  json VStartedStatement::toJson() {
    json funJson = fun->toJson();
    json conJson = {};
    if (con != nullptr) conJson = con->toJson();
    json vStartedStmtJson = {
      {"ntype", "VStartedStatementExpr"}, {"fun", funJson}, {"con", conJson}
    };
    return vStartedStmtJson;
  }

  VRevertedStatement::VRevertedStatement(VFunctionID *_fun, VConstraintExpr *_con) {
    fun = _fun;
    con = _con;
  }
  json VRevertedStatement::toJson() {
    json funJson = fun->toJson();
    json conJson = {};
    if (con != nullptr) conJson = con->toJson();
    json vRevertedStmtJson = {
      {"ntype", "VRevertedStatementExpr"}, {"fun", funJson}, {"con", conJson}
    };
    return vRevertedStmtJson;
  }

  VWillSucceedStatement::VWillSucceedStatement(VFunctionID *_fun, VConstraintExpr *_con) {
    fun = _fun;
    con = _con;
  }
  json VWillSucceedStatement::toJson() {
    json funJson = fun->toJson();
    json conJson = {};
    if (con != nullptr) conJson = con->toJson();
    json vWillSucceedStmtJson = {
      {"ntype", "VWillSucceedStatementExpr"}, {"fun", funJson}, {"con", conJson}
    };
    return vWillSucceedStmtJson;
  }

  VFunctionID::VFunctionID(VConstraintExpr *_base, VID *_fnName, VArgList *_args) {
    base = _base;
    fnName = _fnName;
    args = _args;
  }

  json VFunctionID::toJson() {
    json baseJson = base->toJson();
    json funcJson = fnName->toJson();
    json argsJson = {};
    if (args != nullptr) argsJson = args->toJson();
    json vFunctionIDJson = {
      {"ntype", "VFunctionID"}, {"base", baseJson}, {"func", funcJson}, {"args", argsJson}
    };
    return vFunctionIDJson;
  }

  VArgList::VArgList(vector<VConstraintExpr*> _args) {
    args = _args;
  }
  json VArgList::toJson() {
    vector<json> argsJson;
    for (VConstraintExpr* a : args) {
      argsJson.push_back(a->toJson());
    }
    json vArgListJson = {
      {"ntype", "VArgList"}, {"args", argsJson}
    };
    return vArgListJson;
  }

  VBinExpr::VBinExpr(VConstraintExpr *_lhs, VConstraintExpr *_rhs, VBinOp *_op) {
    lhs = _lhs;
    rhs = _rhs;
    op = _op;
  }
  json VBinExpr::toJson() {
    json lhsJson = lhs->toJson();
    json rhsJson = rhs->toJson();
    json vBinExprJson = {
      {"ntype", "VBinExpr"}, {"lhs", lhsJson}, {"rhs", rhsJson}, {"op", op->op}
    };
    return vBinExprJson;
  }

  VUnExpr::VUnExpr(VConstraintExpr *_expr, VUnOp *_op) {
    expr = _expr;
    op = _op;
  }
  json VUnExpr::toJson() {
    json exprJson = expr->toJson();
    json vUnExprJson = {
      {"ntype", "VUnExpr"}, {"expr", exprJson}, {"op", op->op}
    };
    return vUnExprJson;
  }

  VVarExpr::VVarExpr(VID *_var) {
    var = _var;
  }
  json VVarExpr::toJson() {
    json varJson = var->toJson();
    json vVarExprJson = {
      {"ntype", "VVarExpr"}, {"var", varJson}
    };
    return vVarExprJson;
  }

  VConstExpr::VConstExpr(string _type, string _val) {
    val = _val;
    type = _type;
  }
  json VConstExpr::toJson() {
    return {{"ntype", "VConstExpr"}, {"litType", type}, {"val", val}};
  }

  VFieldAccessExpr::VFieldAccessExpr(VConstraintExpr *_expr, VID *_field) {
    expr = _expr;
    field = _field;
  }
  json VFieldAccessExpr::toJson() {
    json exprJson = expr->toJson();
    json fieldJson = field->toJson();
    json vFieldAccessExprJson = {
      {"ntype", "VFieldAccessExpr"}, {"expr", exprJson}, {"field", fieldJson}
    };
    return vFieldAccessExprJson;
  }

  VArrAccessExpr::VArrAccessExpr(VConstraintExpr *_arr, VConstraintExpr *_idx) {
    arr = _arr;
    idx = _idx;
  }
  json VArrAccessExpr::toJson() {
    json arrJson = arr->toJson();
    json idxJson = idx->toJson();
    json vArrAccessExprJson = {
      {"ntype", "VArrAccessExpr"}, {"arr", arrJson}, {"idx", idxJson}
    };
    return vArrAccessExprJson;
  }

  VFuncCallExpr::VFuncCallExpr(VConstraintExpr *_base, VID *_fnName, VArgList *_args) {
    base = _base;
    fnName = _fnName;
    args = _args;
  }
  json VFuncCallExpr::toJson() {
    json baseJson = {};
    if(base != nullptr) {
        baseJson = base->toJson();
    }
    json funcJson = fnName->toJson();
    json argsJson = args->toJson();
    json vFuncCallExprJson = {
      {"ntype", "VFuncCallExpr"}, {"base", baseJson}, {"func", funcJson}, {"args", argsJson}
    };
    return vFuncCallExprJson;
  }

  VFSumExpr::VFSumExpr(VFunctionID *_func, VConstraintExpr* _arg, VConstraintExpr *_con) {
    func = _func;
    arg = _arg;
    con = _con;
  }
  json VFSumExpr::toJson() {
    json funcJson = func->toJson();
    json argJson = arg->toJson();
    json conJson = con->toJson();
    json vFSumExprJson = {
      {"ntype", "VFSumExpr"}, {"func", funcJson}, {"arg", argJson}, {"con", conJson}
    };
    return vFSumExprJson;
  }

}

File: vpython/vast/vast.py
from typing import List

class VAST():
    pass

class VStatementExpr(VAST):
    pass

class VID(VAST):
    def __init__(self, name: str):
        self.name = name

class VType(VAST):
    def __init__(self, name: str, is_arr: bool):
        self.name = name
        self.is_arr = is_arr

class VVarDecl(VAST):
    def __init__(self, typ: VType, var: VID):
        self.typ = typ
        self.var = var

class VVarDeclList(VAST):
    def __init__(self, var_decs: List[VVarDecl]):
        self.var_decs = var_decs

class VImport(VAST):
    def __init__(self, path: str):
        self.path = path

class VImportList(VAST):
    def __init__(self, imports: List[VImport]):
        self.imports = imports

class VBehavioralSpec(VAST):
    def __init__(self, imports: VImportList, var_decs: VVarDeclList, pre: VStatementExpr, post: VStatementExpr):
        self.imports = imports
        self.var_decs = var_decs
        self.pre = pre
        self.post = post

class VTestSpec(VAST):
    def __init__(self, imports: VImportList, var_decs: VVarDeclList, init: VStatementExpr, spec: VStatementExpr):
        self.imports = imports
        self.var_decs = var_decs
        self.init = init
        self.spec = spec

class VTempSpec(VAST):
    def __init__(self, imports: VImportList, var_decs: VVarDeclList, fairness: VStatementExpr, spec: VStatementExpr):
        self.imports = imports
        self.var_decs = var_decs
        self.fairness = fairness
        self.spec = spec

class VInvSpec(VAST):
    def __init__(self, imports: VImportList, var_decs: VVarDeclList, inv: VStatementExpr):
        self.imports = imports
        self.var_decs = var_decs
        self.inv = inv

class VSynthSpec(VAST):
    def __init__(self, imports: VImportList, var_decs: VVarDeclList, init: VStatementExpr, synth: VStatementExpr):
        self.imports = imports
        self.var_decs = var_decs
        self.init = init
        self.synth = synth

class VBinOp(VAST):
    def __init__(self, op: str):
        self.op = op

class VBinStatementExpr(VStatementExpr):
    def __init__(self, lhs: VStatementExpr, rhs: VStatementExpr, op: VBinOp):
        self.lhs = lhs
        self.rhs = rhs
        self.op = op

class VUnOp(VAST):
    def __init__(self, op: str):
        self.op = op

class VUnStatementExpr(VStatementExpr):
    def __init__(self, con: VStatementExpr, op: VUnOp):
        self.con = con
        self.op = op

class VConstraintExpr(VAST):
    pass

class VArgList(VAST):
    def __init__(self, args: List[VConstraintExpr]):
        self.args = args

    def __len__(self):
        return len(self.args)

    def __iter__(self):
        return self.args.__iter__()

class VFunctionID(VAST):
    def __init__(self, base: VConstraintExpr, fnName: VID, args: VArgList):
        self.base = base
        self.fnName = fnName
        self.args = args

class VExecutedStatement(VStatementExpr):
    def __init__(self, fun: VFunctionID, pre: VConstraintExpr, con: VConstraintExpr):
        self.fun = fun
        self.pre = pre
        self.con = con

class VFinishedStatement(VStatementExpr):
    def __init__(self, fun: VFunctionID, pre: VConstraintExpr, con: VConstraintExpr):
        self.fun = fun
        self.pre = pre
        self.con = con

class VStartedStatement(VStatementExpr):
    def __init__(self, fun: VFunctionID, con: VConstraintExpr):
        self.fun = fun
        self.con = con

class VRevertedStatement(VStatementExpr):
    def __init__(self, fun: VFunctionID, con: VConstraintExpr):
        self.fun = fun
        self.con = con

class VWillSucceedStatement(VStatementExpr):
    def __init__(self, fun: VFunctionID, con: VConstraintExpr):
        self.fun = fun
        self.con = con

class VBinExpr(VConstraintExpr):
    def __init__(self, lhs: VConstraintExpr, rhs: VConstraintExpr, op: VBinOp):
        self.lhs = lhs
        self.rhs = rhs
        self.op = op

class VUnExpr(VConstraintExpr):
    def __init__(self, expr: VConstraintExpr, op: VBinOp):
        self.expr = expr
        self.op = op

class VVarExpr(VConstraintExpr):
    def __init__(self, var: VID):
        self.var = var

class VConstExpr(VConstraintExpr):
    def __init__(self, type: str, val: str):
        self.type = type
        self.val = val

class VFieldAccessExpr(VConstraintExpr):
    def __init__(self, expr: VConstraintExpr, field: VID):
        self.expr = expr
        self.field = field

class VArrAccessExpr(VConstraintExpr):
    def __init__(self, arr: VConstraintExpr, idx: VConstraintExpr):
        self.arr = arr
        self.idx = idx

class VFunctionCallExpr(VConstraintExpr):
    def __init__(self, base: VConstraintExpr, fnName: VID, args: VArgList):
        self.base = base
        self.fnName = fnName
        self.args = args

class VFSumExpr(VConstraintExpr):
    def __init__(self, func: VFunctionID, arg: VConstraintExpr, con: VConstraintExpr):
        self.func = func
        self.arg = arg
        self.con = con
